### PR TITLE
[BEAM-596] Replace DirectResult#awaitCompletion with waitUntilFinish

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This is the PipelineResult interface method, rather than the
DirectRunner-specific method.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @peihe

**Comment by peihe** [comment]
LGTM

**Comment by tgroh** [comment]
R: @lukecwik @bjchambers @dhalperi (Any of) for committer.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java [hunk 1]
```diff
@@ -42,7 +42,7 @@
       "If the pipeline should block awaiting completion of the pipeline. If set to true, "
           + "a call to Pipeline#run() will block until all PTransforms are complete. Otherwise, "
           + "the Pipeline will execute asynchronously. If set to false, the completion of the "
-          + "pipeline can be awaited on by use of DirectPipelineResult#awaitCompletion().")
+          + "pipeline can be awaited on by use of DirectPipelineResult#waitUntilFinish().")
```
Comment (bjchambers) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java]: `PipelineResult#waitUntilFinish()`?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java [hunk 2]
```diff
@@ -390,17 +390,23 @@ public String toString() {
...
-          Thread.currentThread().interrupt();
-          throw e;
-        } catch (Throwable t) {
-          state = State.FAILED;
-          throw t;
+        } catch (Throwable e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          if (e instanceof RuntimeException) {
```
Comment (bjchambers) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: `throw Throwables.propagate(e);`

Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: Do not use -- deprecated.

Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: https://github.com/google/guava/issues/1366

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java (hunk 1 above)
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java]: "can be awaited on" ? Consider rewording for clarity.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java (hunk 2 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: ACK.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectOptions.java]: Done.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java (hunk 2 above)
Comment (bjchambers) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: As per the `@deprecated` note use `throwIfUnchecked`:
https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Throwables.java#L232

Or you can do:

```
Throwables.propagateIfInstanceOf(RuntimeException.class, e);
Throwables.propagateIfInstanceOf(Error.class, e);
throw new RuntimeException(e);
```

**Comment by tgroh** [comment]
I've cleaned this up some. We shouldn't really be catching `Errors` and fiddling with them, so everything now throws an exception.

@bjchambers @dhalperi PTAL

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java [hunk 3]
```diff
@@ -121,12 +121,12 @@ public void run() {
       processElements(evaluator, metricsContainer, enforcements);
 
       finishBundle(evaluator, metricsContainer, enforcements);
-    } catch (Throwable t) {
-      onComplete.handleThrowable(inputBundle, t);
-      if (t instanceof RuntimeException) {
-        throw (RuntimeException) t;
+    } catch (Exception e) {
+      onComplete.handleException(inputBundle, e);
+      if (e instanceof RuntimeException) {
```
Comment (bjchambers) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java]: `Throwables.propagateIfInstanceOf(RuntimeException.class, e);`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java [hunk 4]
```diff
@@ -402,17 +402,20 @@ public MetricResults metrics() {
...
-          Thread.currentThread().interrupt();
-          throw e;
-        } catch (Throwable t) {
-          state = State.FAILED;
-          throw t;
+        } catch (Exception e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          if (e instanceof RuntimeException) {
```
Comment (bjchambers) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: `Throwables.propagateIfInstanceOf`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java (hunk 3 above)
Comment (lukecwik) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java]: I think its clean enough without the Throwables.propagateIfInstanceOf

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java (hunk 4 above)
Comment (lukecwik) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java]: I think its clean enough without the Throwables.propagateIfInstanceOf

**Comment by lukecwik** [comment]
LGTM, merging

**Pull request closed without merging**